pkg/provider/kubernetes/cluster_template_instance.go

```diff
@@ -0,0 +1,171 @@
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	ctrlruntimeclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const ClusterTemplateLabelKey = "template-id"
+
+// AlertmanagerProvider struct that holds required components in order to manage alertmanager objects.
+type ClusterTemplateInstanceProvider struct {
+	// createSeedImpersonatedClient is used as a ground for impersonation
+	// whenever a connection to Seed API server is required
+	createSeedImpersonatedClient impersonationClient
+
+	//privilegedClient is used for admins
+	privilegedClient ctrlruntimeclient.Client
+}
+
+// ClusterTemplateInstanceProvider returns provider
+func NewClusterTemplateInstanceProvider(createSeedImpersonatedClient impersonationClient, privilegedClient ctrlruntimeclient.Client) *ClusterTemplateInstanceProvider {
+	return &ClusterTemplateInstanceProvider{
+		createSeedImpersonatedClient: createSeedImpersonatedClient,
+		privilegedClient:             privilegedClient,
+	}
+}
+
+func ClusterTemplateInstanceProviderFactory(mapper meta.RESTMapper, seedKubeconfigGetter provider.SeedKubeconfigGetter) provider.ClusterTemplateInstanceProviderGetter {
+	return func(seed *kubermaticv1.Seed) (provider.ClusterTemplateInstanceProvider, error) {
+		cfg, err := seedKubeconfigGetter(seed)
+		if err != nil {
+			return nil, err
+		}
+		defaultImpersonationClientForSeed := NewImpersonationClient(cfg, mapper)
+		privilegedClient, err := ctrlruntimeclient.New(cfg, ctrlruntimeclient.Options{Mapper: mapper})
+		if err != nil {
+			return nil, err
+		}
+		return NewClusterTemplateInstanceProvider(
+			defaultImpersonationClientForSeed.CreateImpersonatedClient,
+			privilegedClient,
+		), nil
+	}
+}
+
+func (r ClusterTemplateInstanceProvider) Create(userInfo *provider.UserInfo, template *kubermaticv1.ClusterTemplate, project *kubermaticv1.Project, replicas int64) (*kubermaticv1.ClusterTemplateInstance, error) {
+	impersonationClient, err := createImpersonationClientWrapperFromUserInfo(userInfo, r.createSeedImpersonatedClient)
+	if err != nil {
+		return nil, err
+	}
+	instance := createClusterTemplateInstance(template, project, replicas)
+
+	err = impersonationClient.Create(context.Background(), instance)
+	return instance, err
+}
+
+func (r ClusterTemplateInstanceProvider) CreateUnsecured(template *kubermaticv1.ClusterTemplate, project *kubermaticv1.Project, replicas int64) (*kubermaticv1.ClusterTemplateInstance, error) {
+
+	instance := createClusterTemplateInstance(template, project, replicas)
+
+	err := r.privilegedClient.Create(context.Background(), instance)
+	return instance, err
+}
+
+func createClusterTemplateInstance(template *kubermaticv1.ClusterTemplate, project *kubermaticv1.Project, replicas int64) *kubermaticv1.ClusterTemplateInstance {
+	instance := &kubermaticv1.ClusterTemplateInstance{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:   GetClusterTemplateInstanceName(project.Name, template.Name),
```

Comment:
user can create multiple ClusterTemplateInstances on the same project with the same template right? Then this wont be unique anymore.

Reply:
in my opinion, should create only one. Doesn't make sense to create a different group of clusters because clusters will be the same from the same template. This solution will allow creating group of the clusters from the template which can be scale up/down

Reply:
right, so each create will act like an update and the only thing that can change is the replica count

Reply:
exactly